### PR TITLE
options for fixing broken paragraph spacing

### PR DESCRIPTION
**Pull request by HazelSh**
So this is the reworked fix for the broken paragraph spacing in some royalroad stories, see #690.

There's two fixes for two issues. 

First, some stories have empty `<p></p>` tags, causing too much spacing between paragraphs. Noticed this most recently in Only Villians Do That, but I've seen it in other stories too. 

Second, other stories have certain chapters where the spacing between paragraphs is totally removed. Looking into it, this is because those chapters are formatted with `<div>` tags not `<p>` tags. As examples, see ch. 25 of The Flower That Bloomed Nowhere, and ch. 33 of Wizard's Tower. Looks fine on the site, broken paragraph spacing on kindle through calibre and fanficfare.

These stories are fixed with this patch, and this patch includes default-enabling the new settings for Royal Road. Haven't actually checked yet that this doesn't massively break other things on that site, but I *think* it'll be fine.

### CODE REVIEW

**Review by JimmXinu — COMMENTED** [review]
I'm not very inclined to include these changes, but I'm willing to discuss it.  Details below.

FYI, please include story URLs, not just story titles.  There are, for example, two stories titled 'Wizard's Tower'

(I tweaked your post to put `code` ticks around your p & div tags because they were disappearing.)

File: fanficfare/adapters/base_adapter.py [hunk 1]
```diff
@@ -726,6 +726,19 @@ def _do_utf8FromSoup(self,url,soup,fetch=None,allow_replace_br_with_p=True):
...
 
+        # replaces <div> tags with <p> tags.
+        # adds proper paragraph spacing to some chapters on some royalroad stories (as of early 2022)
+        if self.getConfig("div_to_p", default=False):
+            for tag in soup.find_all('div'):
+                tag.name = 'p'
+        
+        # strips empty <p> </p> tags that do nothing but add excessive whitespace
+        # issue exists on some royalroad stories as of mid 2021
+        if self.getConfig("remove_empty_p", default=False):
```
[the remark this answers is not in the document]
Comment (HazelSh) [reply fanficfare/adapters/base_adapter.py:737]: yeah, ok. I missed that feature. taking <code>p</code> out of keep_empty_tags (defaults.ini line 400) does it. I'm kind of curious to know why it is in there, now.

File: fanficfare/adapters/base_adapter.py [hunk 2]
```diff
@@ -726,6 +726,19 @@ def _do_utf8FromSoup(self,url,soup,fetch=None,allow_replace_br_with_p=True):
             if "%s"%ae != "'NoneType' object has no attribute 'next_element'":
                 logger.error("Error parsing HTML, probably poor input HTML. %s"%ae)
 
+        # replaces <div> tags with <p> tags.
+        # adds proper paragraph spacing to some chapters on some royalroad stories (as of early 2022)
+        if self.getConfig("div_to_p", default=False):
```
[the remark this answers is not in the document]
Comment (HazelSh) [reply fanficfare/adapters/base_adapter.py:731]: ok, yeah. bad idea. it was already breaking things, like the author's notes.

File: fanficfare/adapters/adapter_fictionlive.py [hunk 3]
```diff
@@ -114,7 +114,7 @@ def extract_metadata(self, data, get_cover):
         # cht is *chunktime* -- newest story chunk added.
         # ut for update time includes other kinds of update -- threads, chat etc
         # ct <= rt <= cht <= ut
-        self.story.setMetadata("dateUpdated", self.parse_timestamp(data['cht']))
+        self.story.setMetadata("dateUpdated", self.parse_timestamp(data['cht'] if 'cht' in data else data['ut']))
```
[the remark this answers is not in the document]
Comment (HazelSh) [reply fanficfare/adapters/adapter_fictionlive.py:117]: ah, yes. that's unrelated and will come with a minor fictionlive update at some point

**Comment by HazelSh** [comment]
yeah, ok. bad ideas all round. let me think about this a bit

**Pull request closed without merging**